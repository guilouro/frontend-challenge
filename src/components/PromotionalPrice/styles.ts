import styled, { css, DefaultTheme } from 'styled-components'
import { PromotionalPriceProps } from '.'

type PriceProps = {
  isPromotional?: boolean
} & Pick<PromotionalPriceProps, 'dark'>

const priceModifiers = {
  promotional: (theme: DefaultTheme) => css`
    color: ${theme.colors.textLight};
    text-decoration: line-through;
    margin-right: ${theme.spacings.xsmall};
    font-weight: ${theme.font.normal};
  `,

  dark: (theme: DefaultTheme) => css`
    color: ${theme.colors.text};
  `
}

export const Price = styled.div<PriceProps>`
  ${({ theme, dark, isPromotional = false }) => css`
    display: inline-flex;
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.base};

    ${isPromotional && priceModifiers.promotional(theme)}
    ${dark && priceModifiers.dark(theme)}
  `}
`
