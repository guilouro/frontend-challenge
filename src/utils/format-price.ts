export default function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR'
  }).format(price)
}
