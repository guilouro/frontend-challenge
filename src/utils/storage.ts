import { CartObject } from 'hooks/use-cart'

const APP_KEY = 'BARKYN_CART'

export function getStorageItem() {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(APP_KEY)
  return JSON.parse(data || '[]')
}

export function setStorageItem(value: CartObject) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(APP_KEY, data)
}
