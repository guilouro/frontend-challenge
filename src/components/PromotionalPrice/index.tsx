import formatPrice from 'utils/format-price'
import * as S from './styles'

export type PromotionalPriceProps = {
  price: number
  promotionalPrice?: number
  dark?: boolean
}

const PromotionalPrice = ({
  price,
  promotionalPrice = 0,
  dark = false
}: PromotionalPriceProps) => (
  <div>
    {promotionalPrice ? (
      <S.Price isPromotional>{formatPrice(price)}</S.Price>
    ) : null}
    <S.Price dark={dark}>{formatPrice(promotionalPrice || price)}</S.Price>
  </div>
)

export default PromotionalPrice
