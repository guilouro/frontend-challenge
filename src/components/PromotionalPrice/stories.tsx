import { Story, Meta } from '@storybook/react'
import PromotionalPrice, { PromotionalPriceProps } from '.'

export default {
  title: 'PromotionalPrice',
  component: PromotionalPrice,
  args: {
    promotionalPrice: 1.99,
    price: 2.98,
    dark: false
  }
} as Meta

export const Default: Story<PromotionalPriceProps> = (args) => (
  <PromotionalPrice {...args} />
)
