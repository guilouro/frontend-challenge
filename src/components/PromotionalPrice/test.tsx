import 'intl.mock'
import { render, screen } from 'utils/test-utils'

import PromotionalPrice from '.'

describe('<PromotionalPrice />', () => {
  it('should render price in label', () => {
    render(<PromotionalPrice price={2.29} />)

    expect(screen.getByText('2.29')).not.toHaveStyle({
      textDecoration: 'line-through'
    })
  })

  it('should render a line-through in price when promotional', () => {
    const { container } = render(
      <PromotionalPrice price={2.29} promotionalPrice={1.29} />
    )

    expect(screen.getByText('2.29')).toHaveStyle({
      textDecoration: 'line-through'
    })

    expect(screen.getByText('1.29')).not.toHaveStyle({
      textDecoration: 'line-through'
    })

    expect(container.firstChild).toMatchSnapshot()
  })
})
