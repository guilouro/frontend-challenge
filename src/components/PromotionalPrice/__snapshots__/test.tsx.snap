// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<PromotionalPrice /> should render a line-through in price when promotional 1`
] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  font-weight: 600;
  color: #5f2eff;
  color: #6c6b76;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  margin-right: 1.6rem;
  font-weight: 400;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  font-weight: 600;
  color: #5f2eff;
}

<div>
  <div
    class="c0"
  >
    2.29
  </div>
  <div
    class="c1"
  >
    1.29
  </div>
</div>
`
